File: backend/internal/models/listing_menu.go
package models

import "time"

// ListingMenu is one banquet-menu variant belonging to a Listing —
// optionally scoped to one specific ListingHall (HallID nil means "this
// menu applies venue-wide, regardless of which hall ends up booked").
// Like ListingHall, deliberately no back-reference slice association on
// Listing/ListingHall — see listing_hall.go's own doc comment on why; the
// full menu tree is only ever assembled by an explicit handler-level
// query (listing_handler.go's Get/Menus), never a Preload.
type ListingMenu struct {
	ID            uint   `gorm:"primaryKey" json:"id"`
	ListingID     uint   `gorm:"not null;index" json:"listing_id"`
	HallID        *uint  `gorm:"index" json:"hall_id,omitempty"`
	NameRu        string `gorm:"size:200;not null" json:"name_ru"`
	NameKz        string `gorm:"size:200;not null" json:"name_kz"`
	DescriptionRu string `gorm:"type:text" json:"description_ru,omitempty"`
	DescriptionKz string `gorm:"type:text" json:"description_kz,omitempty"`
	PricePerGuest uint   `gorm:"not null;default:0" json:"price_per_guest"`
	// MinGuests/MaxGuests here are this specific menu's own order-size
	// constraints (e.g. "only bookable for 50+ guests") — a third,
	// independent concept from Listing.MinGuests/MaxGuests (pricing-tier
	// range) and ListingHall.Capacity (physical capacity). Pointers, not
	// plain uint, because 0 is not a sensible "no constraint" default the
	// way it is for a capacity/price field — nil unambiguously means "not
	// set" for consistency with Event.EventDate/EventRequest.SubmittedAt's
	// own existing *time.Time convention for optional scalars in this
	// codebase.
	MinGuests *uint `json:"min_guests,omitempty"`
	MaxGuests *uint `json:"max_guests,omitempty"`
	// IsActive — see ListingHall.IsActive's own doc comment on why this
	// carries no `default:` tag: GORM silently drops an explicit false
	// for a bool field tagged with a default (treating it as "unset"),
	// which is exactly what "Дублировать меню" (draft/inactive copies)
	// needs to actually work. Pure Go struct-tag fix, no migration
	// needed — verified directly against this Postgres/GORM version.
	IsActive   bool       `gorm:"not null" json:"is_active"`
	SortOrder  int        `gorm:"not null;default:0" json:"sort_order"`
	ValidFrom  *time.Time `json:"valid_from,omitempty"`
	ValidUntil *time.Time `json:"valid_until,omitempty"`
	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  time.Time  `json:"updated_at"`
}

// ListingMenuSection groups ListingMenuItems under a heading (e.g.
// "Салаты", "Горячее") within one ListingMenu.
type ListingMenuSection struct {
	ID        uint   `gorm:"primaryKey" json:"id"`
	MenuID    uint   `gorm:"not null;index" json:"menu_id"`
	TitleRu   string `gorm:"size:150;not null" json:"title_ru"`
	TitleKz   string `gorm:"size:150;not null" json:"title_kz"`
	SortOrder int    `gorm:"not null;default:0" json:"sort_order"`
}

// ListingMenuItem is one dish within a ListingMenuSection.
type ListingMenuItem struct {
	ID            uint   `gorm:"primaryKey" json:"id"`
	SectionID     uint   `gorm:"not null;index" json:"section_id"`
	NameRu        string `gorm:"size:200;not null" json:"name_ru"`
	NameKz        string `gorm:"size:200;not null" json:"name_kz"`
	DescriptionRu string `gorm:"type:text" json:"description_ru,omitempty"`
	DescriptionKz string `gorm:"type:text" json:"description_kz,omitempty"`
	// QuantityText is free-form ("200 г", "1 порция на гостя") rather than
	// a structured amount+unit pair — matches how little structure this
	// codebase already imposes on similar display-only text elsewhere
	// (e.g. Listing.DescriptionRu/Kz).
	QuantityText string `gorm:"size:100" json:"quantity_text,omitempty"`
	SortOrder    int    `gorm:"not null;default:0" json:"sort_order"`
}

// ListingMenuExtra is an optional add-on priced independently of the
// per-guest menu price (e.g. an extra drink package, a service fee, decor)
// — Type/Unit are free-form strings rather than DB enums, matching how
// Event.Type/Booking.Status are already modeled in this codebase.
type ListingMenuExtra struct {
	ID            uint   `gorm:"primaryKey" json:"id"`
	MenuID        uint   `gorm:"not null;index" json:"menu_id"`
	Type          string `gorm:"size:30;not null" json:"type"`
	TitleRu       string `gorm:"size:200;not null" json:"title_ru"`
	TitleKz       string `gorm:"size:200;not null" json:"title_kz"`
	Price         uint   `gorm:"not null;default:0" json:"price"`
	Unit          string `gorm:"size:30" json:"unit,omitempty"`
	DescriptionRu string `gorm:"type:text" json:"description_ru,omitempty"`
	DescriptionKz string `gorm:"type:text" json:"description_kz,omitempty"`
	SortOrder     int    `gorm:"not null;default:0" json:"sort_order"`
}

File: flutter/lib/models/listing_menu.dart
import '../state/locale_provider.dart';
import 'listing_menu_extra.dart';
import 'listing_menu_section.dart';

/// Mirrors backend/internal/models/listing_menu.go's `ListingMenu` plus the
/// `sections`/`extras` arrays `menuOut` (listing_handler.go) nests onto it
/// wherever the full tree is loaded (`GET /api/listings/:id` and
/// `GET /api/listings/:id/menus`) — both default to empty so a plain
/// `ListingMenu` (e.g. from the admin CRUD response, which returns the bare
/// row) still parses.
class ListingMenu {
  const ListingMenu({
    required this.id,
    required this.listingId,
    this.hallId,
    required this.nameRu,
    required this.nameKz,
    this.descriptionRu = '',
    this.descriptionKz = '',
    required this.pricePerGuest,
    this.minGuests,
    this.maxGuests,
    required this.isActive,
    required this.sortOrder,
    this.validFrom,
    this.validUntil,
    this.sections = const [],
    this.extras = const [],
  });

  final int id;
  final int listingId;

  /// null = "available for any hall" — never assumed to mean "no hall
  /// exists"; see models.ListingMenu's own doc comment.
  final int? hallId;
  final String nameRu;
  final String nameKz;
  final String descriptionRu;
  final String descriptionKz;
  final int pricePerGuest;

  /// This menu's own order-size constraints — independent of
  /// `Listing.minGuests/maxGuests` (a pricing-tier range) and
  /// `ListingHall.capacity` (physical capacity). Pointers on the backend
  /// (0 is not a valid "no constraint" default), so nullable here too —
  /// never coerced to 0.
  final int? minGuests;
  final int? maxGuests;
  final bool isActive;
  final int sortOrder;
  final DateTime? validFrom;
  final DateTime? validUntil;
  final List<ListingMenuSection> sections;
  final List<ListingMenuExtra> extras;

  factory ListingMenu.fromJson(Map<String, dynamic> json) {
    return ListingMenu(
      id: json['id'] as int,
      listingId: json['listing_id'] as int? ?? 0,
      hallId: json['hall_id'] as int?,
      nameRu: json['name_ru'] as String? ?? '',
      nameKz: json['name_kz'] as String? ?? '',
      descriptionRu: json['description_ru'] as String? ?? '',
      descriptionKz: json['description_kz'] as String? ?? '',
      pricePerGuest: json['price_per_guest'] as int? ?? 0,
      minGuests: json['min_guests'] as int?,
      maxGuests: json['max_guests'] as int?,
      isActive: json['is_active'] as bool? ?? true,
      sortOrder: json['sort_order'] as int? ?? 0,
      validFrom: json['valid_from'] != null
          ? DateTime.tryParse(json['valid_from'] as String)
          : null,
      validUntil: json['valid_until'] != null
          ? DateTime.tryParse(json['valid_until'] as String)
          : null,
      sections:
          (json['sections'] as List?)
              ?.map(
                (e) => ListingMenuSection.fromJson(
                  Map<String, dynamic>.from(e as Map),
                ),
              )
              .toList() ??
          const [],
      extras:
          (json['extras'] as List?)
              ?.map(
                (e) => ListingMenuExtra.fromJson(
                  Map<String, dynamic>.from(e as Map),
                ),
              )
              .toList() ??
          const [],
    );
  }

  String name(AppLocale locale) => t(locale, ru: nameRu, kz: nameKz);
  String description(AppLocale locale) =>
      t(locale, ru: descriptionRu, kz: descriptionKz);
}
